From 33663cdb2901b70d47a19970a9cfaf206f7743de Mon Sep 17 00:00:00 2001
From: Jin Yanjiang <yanjiang.jin@windriver.com>
Date: Wed, 14 Sep 2011 15:57:44 +0800
Subject: [PATCH] mips/nlm_xlp_64_be: enable NAE in the 2nd kernel booted by kexec

If you boot a 2nd kernel via kexec it will hang if the frequency is
set again. To fix this issue, we should not call the function
set_nae_frequency() on booting in the 2nd kernel.
The function set_nae_frequency() will set the net field of
SYS_DFS_DIV_DEC_CTRL to 1 for NAE.
As the Programming Reference Manual of XLP Processor said:
"Setting a read/write field to 1 will increment the DFS clock divider for
a device or interface by one."
The system clock of net was set by the first kernel, and was not cleared
after the secondary kernel booted up.
So if the secondary kernel set it again, the NAE frequency will double, and
then system will hang.

Signed-off-by: Jin Yanjiang <yanjiang.jin@windriver.com>
---
 arch/mips/include/asm/kexec.h                   |    1 +
 arch/mips/include/asm/mach-netlogic/nlm_kexec.h |    1 -
 arch/mips/netlogic/common/nlm_hal_nae.c         |   11 +++++++----
 drivers/net/nae/xlp_nae.c                       |    6 ------
 4 files changed, 8 insertions(+), 11 deletions(-)

diff --git a/arch/mips/include/asm/kexec.h b/arch/mips/include/asm/kexec.h
index b282f83..63b181c 100644
--- a/arch/mips/include/asm/kexec.h
+++ b/arch/mips/include/asm/kexec.h
@@ -10,6 +10,7 @@
 #define _MIPS_KEXEC
 
 #include <asm/stacktrace.h>
+#include <asm/string.h>
 
 /* Maximum physical address we can use pages from */
 #define KEXEC_SOURCE_MEMORY_LIMIT (0x20000000)
diff --git a/arch/mips/include/asm/mach-netlogic/nlm_kexec.h b/arch/mips/include/asm/mach-netlogic/nlm_kexec.h
index 035a131..4e3e7c0 100644
--- a/arch/mips/include/asm/mach-netlogic/nlm_kexec.h
+++ b/arch/mips/include/asm/mach-netlogic/nlm_kexec.h
@@ -23,7 +23,6 @@
 #include <linux/ctype.h>
 
 #include <asm/netlogic/hal/nlm_hal_nae.h>
-#include <asm/mach-netlogic/nlm_kexec.h>
 #include <linux/bootmem.h>
 #include <asm/setup.h>
 
diff --git a/arch/mips/netlogic/common/nlm_hal_nae.c b/arch/mips/netlogic/common/nlm_hal_nae.c
index 9649d9e..37b439f 100644
--- a/arch/mips/netlogic/common/nlm_hal_nae.c
+++ b/arch/mips/netlogic/common/nlm_hal_nae.c
@@ -22,6 +22,7 @@ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 THE POSSIBILITY OF SUCH DAMAGE.
 *****************************#NETL_2#********************************/
+#include <asm/mach-netlogic/nlm_kexec.h>
 
 #include "nlm_hal_fmn.h"
 #include "nlm_hal_nae.h"
@@ -246,7 +247,7 @@ int rely_on_firmware_config = 0;
 #define GET_NAE_GLOBAL_PROP(prop, buf, len)					\
 	copy_fdt_prop(fdt, NAE_GLOBAL_NODE, prop, PROP_CELL, buf, len)
 
-#define MAX_PROP_LEN 30
+#define NAE_MAX_PROP_LEN 30
 /* Temporarily specifying these sizes here.
    These will be moved to FDT soon
 */
@@ -571,7 +572,7 @@ static void set_nae_frequency(int frequency)
 static void parse_fdt_nae_config(void *fdt)
 {
 	int i = 0, port_node;
-	char port_type_str[MAX_PROP_LEN];
+	char port_type_str[NAE_MAX_PROP_LEN];
 	int size = 0;
 	int num_ports = 0;
 	uint32_t start_port, num_nae_regs, num_intf_regs;
@@ -587,7 +588,9 @@ static void parse_fdt_nae_config(void *fdt)
 
 	if(GET_NAE_PROP("frequency", &frequency, sizeof(uint32_t)) < 0)
 		nlm_print("fdt missing frequency\n");
-	set_nae_frequency(frequency);
+
+	if (!is_kexec_boot())
+		set_nae_frequency(frequency);
 
 	/*******************************************************************************/
 	/* Read range of ports allocated per domain
@@ -710,7 +713,7 @@ static void parse_fdt_nae_config(void *fdt)
 
 		GET_PORT_PROP("num-channels", &nae_cfg.ports[port].num_channels, sizeof(uint32_t));
 	
-		GET_PORT_STR_PROP("type", port_type_str, MAX_PROP_LEN);
+		GET_PORT_STR_PROP("type", port_type_str, NAE_MAX_PROP_LEN);
 		if (!strcmp(port_type_str, "SGMII_IF")) {
 			nae_cfg.ports[port].iftype = SGMII_IF;
 		}
diff --git a/drivers/net/nae/xlp_nae.c b/drivers/net/nae/xlp_nae.c
index d002ed0..a52947d 100644
--- a/drivers/net/nae/xlp_nae.c
+++ b/drivers/net/nae/xlp_nae.c
@@ -54,7 +54,6 @@
 #include <asm/netlogic/hal/nlm_hal_pic.h>
 
 #include <asm/netlogic/hal/nlm_hal_macros.h>
-#include <asm/mach-netlogic/nlm_kexec.h>
 
 #include <asm/netlogic/mips-exts.h>
 
@@ -1108,11 +1107,6 @@ static struct pci_driver soc_driver = {
 
 static int __init nlm_xlp_mac_init(void)
 {
-	if (is_kexec_boot()) {
-		pr_err("%s: NAE driver doesn't work when booting from kexec\n", __func__);
-		return -ENODEV;
-	}
-
 	nlm_xlp_nae_init();
 
 	return pci_register_driver(&soc_driver);
-- 
1.7.0.4

